Merge IOHK fork

peterbecich opened the pull request:
https://github.com/eskimor/purescript-bridge/issues/63


* `CodeGenSwitches` no longer used
* `Proxy` no longer used
* `Printer` uses `Leijen.Text` instead of `Text`
* `genericShow` PureScript instance
  * I think the same as https://github.com/eskimor/purescript-bridge/pull/85
    * PR 85 has been replaced with IOHK's implementation  
* some unit tests replaced with IOHK's tests
* all printer imports are combined into one function `instanceToImportLines` instead of being divided between two modules
* excellent `RoundTrip` tests implemented by IOHK
  * greatly improves test coverage for both Argonaut and `json-helpers` and exposes issues

TODO
* ~check that derived `Generic` instances are correct~
  * https://github.com/eskimor/purescript-bridge/issues/74#issuecomment-1498529423
* ~fix `RoundTrip` test~
* ~some other broken tests~


------

The PR supports https://github.com/input-output-hk/purescript-bridge-json-helpers and the existing library https://github.com/coot/purescript-argonaut-aeson-generic

The issues with `purescript-argonaut-aeson-generic` are documented in the readme.  Importantly, these issues exist on `master` right now without the PR, so I think it should not block merging.  Discussion https://github.com/purescript-contrib/purescript-argonaut-codecs/issues/115#issuecomment-1863200986


The other library https://github.com/paf31/purescript-foreign-generic continues to be supported, but the test coverage is much less than the other two.

shmish111 committed:
add optional Foreign encoding and decoding

shmish111 committed:
remove TODO

shmish111 committed:
use https://github.com/paf31/purescript-foreign-generic/pull/52

krisajenkins committed:
Whitespace changes.

krisajenkins committed:
Refactoring the code to use a structured printing library.

krisajenkins committed:
Bugfix for Encode/Decode constraints.

For generic decoding, `Decode (Foo a)` needs a `Decode a` constraint,
but it *doesn't* need a `Generic a` constraint.

Adding in this unnecessary constraint creates needless problems like,
"there isn't an instance for `Generic String _`."

krisajenkins committed:
Bugfix for Encode/Decode constraints.

For generic decoding, `Decode (Foo a)` needs a `Decode a` constraint,
but it *doesn't* need a `Generic a` constraint.

Adding in this unnecessary constraint creates needless problems like,
"there isn't an instance for `Generic String _`."

krisajenkins committed:
Bugfix for decode/encode instances on recursive types.

PureScript isn't happy with eta-reduced typeclass instances on recursive
types. We have to make the function application explicit.

krisajenkins committed:
You can now generate `Eq` instances for something like `Proxy :: Proxy (Container A)`.

The generated code will include an `Eq a =>` constraint.

krisajenkins committed:
Whitespace changes.

krisajenkins committed:
You can now generate a `Show` instance based on `genericShow`.

krisajenkins committed:
You can now generate a `Show` instance based on `genericShow`.

krisajenkins committed:
Aeson-compatible encodings for Enums*.

* Sum types where every constructor has zero arguments. Aeson has
special handling for these.

krisajenkins committed:
Bugfix for genericShow.

It turns out you can't eta-reduce typeclass instances for
recursively-defined typeclasses in PureScript. That is:

`show = genericShow`

...has to be replaced with:

`show x = genericShow`

...to work reliably.

See: https://github.com/purescript/purescript/issues/2975

krisajenkins committed:
genericShow now supports parameterised types.

That is, `(genericShow <*> mkSumType) (Proxy @(Foo A))` will generate:

```
instance showFoo :: Show a => Show (Foo a) where
  show = genericShow
```

...whereas before it would have missed out the `Show a` constraint.

krisajenkins committed:
Adding Functor support.

shmish111 committed:
fix haddock

krisajenkins committed:
Applying hlint suggestions.

krisajenkins committed:
You can now generate `Ord` instances for something like `Proxy :: Proxy (Container A)`.

The generated code will include an `Ord a =>` constraint.

krisajenkins committed:
Version 0.13.1.0.

krisajenkins committed:
Updating for PureScript 0.13.

krisajenkins committed:
Adding support for Eq1 instances.

krisajenkins committed:
Factoring out some duplicated code in the typeclass instance-generation.

shmish111 committed:
use a newer stackage

jhbertra committed:
Merge pull request #1 from input-output-hk/eq1

Eq1

Commit:
run lorri init

Commit:
Add flake.nix

Commit:
Update import module for genericShow

Commit:
0.14.0.0

Commit:
Remove shell params

jhbertra committed:
Merge pull request #3 from input-output-hk/jhbertra/ps-014

Update for PureScript 0.14

Commit:
Replace foreign-generic with Argonaut

Commit:
Add roundtrip test project

Commit:
Add compile test

Commit:
Add Aesone round trip test

Commit:
Spawn only one process and check warnings

Commit:
Remove Proxy from API

Commit:
Add lenses to test

Commit:
Fix bugs with nested decoders

Commit:
Simplify test type

Commit:
Integrate json helpers

Commit:
Refactor printing

jhbertra committed:
Merge pull request #4 from input-output-hk/jhbertra/ps-014

PureScript 14 updates, Argonaut encode and decode instances

Commit:
Export everything again

Commit:
Fix problems with inline records in sum types

Commit:
Make sum type encoding better

Commit:
Update sum type encode/decode

Commit:
Add recursive types to test

jhbertra committed:
Merge pull request #5 from input-output-hk/jhbertra/bug-fixes

Bug fixes

Commit:
Fix handling of multi record constructors

Commit:
Fix encoding of nullary constructors

Commit:
Add support for set and map json

Commit:
Update json-helpers dependency

Commit:
Add ormolu to nix shell

Commit:
Add support for custom instances

Commit:
Format everything with ormolu

Commit:
Add newlines at the end of module files

Commit:
Fix type extraction for custom instances

Commit:
Add import lines for custom instances

paluh committed:
Add package namespacing support

paluh committed:
Make package namespacing introduction backward compatible.

jhbertra committed:
Merge pull request #10 from paluh/master

Add package namespacing support

peterbecich committed:
`nix flake update`

latest `haskell-flake`

also uses https://github.com/thomashoneyman/purescript-overlay

peterbecich committed:
copy Stylish Haskell, Nix Flake config

peterbecich committed:
fix compilation

peterbecich committed:
format with Fourmolu and then Stylish Haskell

```
fourmolu -i src --comma-style leading --import-export-style leading --function-arrows leading --indentation 4
--record-brace-space true

fourmolu -i test --comma-style leading --import-export-style leading --function-arrows leading --indentation 4
--record-brace-space true

find . -name '*.hs' | xargs stylish-haskell -i
```

to reduce diffs from parent repository

peterbecich committed:
WIP merge IOHK fork

* `CodeGenSwitches` no longer used
* `Proxy` no longer used
* `Printer` uses `Leijen.Text` instead of `Text`
* some unit tests replaced with IOHK's tests

TODO
* fix derived instances
  * https://github.com/eskimor/purescript-bridge/issues/74#issuecomment-1498529423
* fix tests

peterbecich committed:
`genericShow` instance provided by IOHK fork

delete second implementation of this instance

Merge remote-tracking branch 'parent/master' into merge-iohk-3

peterbecich committed:
Encode/Decode instances by IOHK using `purescript-bridge-json-helpers`

https://github.com/input-output-hk/purescript-bridge-json-helpers

peterbecich committed:
bump to version 0.16.0.0; delete boilerplate

peterbecich committed:
generate `DecodeJsonField`

peterbecich committed:
delete unused `CodeGenSwitches`

`SumTypes` replace `CodeGenSwitches`

peterbecich committed:
fix LSP server for RoundTrip test

peterbecich committed:
fix rendered `EncodeJson` and `DecodeJson` instances

`RoundTrip` builds successfully now:

```
cabal test --test-options="--match \"/writePSTypesWith/should be buildable/\""
```

peterbecich commented:
Current error in `RoundTrip` test

```
cabal test --test-options="--match \"/writePSTypesWith/should produce aeson-compatible argonaut instances/\""
```

```
  3) writePSTypesWith should produce aeson-compatible argonaut instances
       Falsifiable (after 1 test):
         Either (Right (Just False))
       {"contents":{"Right":false},"tag":"Either"}
       expected: ""
        but got: "got {\"contents\":{\"Right\":false},\"tag\":\"Either\"}"
```

peterbecich committed:
extend example to show failing case

```
Error decoding Foo: An error occurred while decoding a JSON value:
  Under 'When decoding a Foo':
  Under '"tag" property is missing':
  No value was found.
```

peterbecich committed:
fix printing of import list

peterbecich committed:
fix printing of imports with/without aliases

import lines with and without aliases should not be merged

peterbecich committed:
RoundTrip test succeeds with json-helper library

peterbecich committed:
RoundTrip tests for both json-helpers and argonaut-aeson-generic

peterbecich committed:
fix dependency notice for `json-helpers`

"The following purescript packages are needed by the generated code:"

peterbecich committed:
clean up example

peterbecich committed:
fix example using argonaut-aeson-generic

peterbecich committed:
improve example to show random errors

see error by:
- enter `example` directory
- spago bundle-app --to static/index.js
- cabal run example
- open http://localhost:8080/index.html in browser
- see browser console
- refresh the page. Sometimes the client will decode the payload successfully,
  sometimes not

peterbecich commented:
Encoding with `argonaut-aeson-generic` has issues. Can be demonstrated by running the `example`. The `example` generates a random payload. It is served to the client. If the client decodes the payload successfully, the client modifies some values and sends the payload back to the server.

- enter `example` directory
- `spago bundle-app --to static/index.js`
- `cabal run example`
- open http://localhost:8080/index.html in browser
- see browser console
- refresh the page. Sometimes the client will decode the payload successfully, sometimes not



Example success, from server:
```
Serving:
{
    "_fooBaz": {
        "_bazMessage": "hello"
    },
    "_fooList": [
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20
    ],
    "_fooMap": {
        "bar": 3,
        "baz": 3,
        "foo": 2
    },
    "_fooMessage": "Hello",
    "_fooNumber": 123,
    "_fooTestData": {
        "contents": {
            "contents": false,
            "tag": "Bool"
        },
        "tag": "Maybe"
    },
    "_fooTestSum": {
        "contents": -5.6,
        "tag": "Number"
    }
}
Received from client:
Foo message: Hola        Foo number: 124         Foo list length: 22     Foo Map length: 4
```

Example failure, from server:

```
Serving:
{
    "_fooBaz": {
        "_bazMessage": "hello"
    },
    "_fooList": [
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20
    ],
    "_fooMap": {
        "bar": 3,
        "baz": 3,
        "foo": 2
    },
    "_fooMessage": "Hello",
    "_fooNumber": 123,
    "_fooTestData": {
        "contents": {
            "Right": true
        },
        "tag": "Either"
    },
    "_fooTestSum": {
        "tag": "Nullary"
    }
}
```
Decoding error message on client:
```
Error decoding Foo: An error occurred while decoding a JSON value:
  Under 'When decoding a Foo':
  At object key '_fooTestData':
  Under 'When decoding a Either':
  Under 'Either':
  At object key 'tag':
  No value was found.
```

I don't know why this `Either` payload is missing its value.

flip111 commented:
@peterbecich can you clarify this?

> Encoding with argonaut-aeson-generic has issues

> Decoding error message on client

Is it both encoding and decoding that are problematic ?

flip111 commented:
When i edit your code

Types.hs
```haskell
data Foo = Foo
  { _fooMessage :: Text
  , _fooNumber  :: Int
  , _fooList    :: [Int]
  , _fooMap     :: Map.Map Text Int
  , _fooBaz     :: Baz
  , _fooTestData :: Either String Bool
  }
  deriving (FromJSON, Generic, ToJSON)
```

MyLib.hs
```haskell
foo :: Foo
foo = Foo
  (pack "Hello")
  123
  [10..20]
  (Map.fromList [(pack "foo", 2), (pack "bar", 3), (pack "baz", 3)])
  (Baz $ pack "hello")
  (Right True)
```

then i do get a missing Either tag, but it seems to be the server who doesn't send it rather than the client making a mistake of not decoding it properly.

```
Hello, Haskell!
Serving Foo:
{
    "_fooBaz": {
        "_bazMessage": "hello"
    },
    "_fooList": [
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20
    ],
    "_fooMap": {
        "bar": 3,
        "baz": 3,
        "foo": 2
    },
    "_fooMessage": "Hello",
    "_fooNumber": 123,
    "_fooTestData": {
        "Right": true
    }
}
```

flip111 commented:
# running the example
```shell
cd example
spago bundle-app --to static/index.js
stack run # (implies latest LTS)
```

Visit http://localhost:8080/index.html

# General remarks

* Would like to update stack/GHC to the latest LTS

peterbecich committed:
fix Stack project

peterbecich committed:
temporarily disable tests in Nix

peterbecich commented:
@flip111 , thanks for taking a look at this.

I have updated the Stack project.

Two alternative ways to run the example:
- in project root, `nix run .#example`. Then http://localhost:8080/index.html should be up
- `cd example` , `cabal run example`, http://localhost:8080/index.html

peterbecich committed:
improve example

peterbecich commented:
This object is encoded correctly server-side, and decoded successfully client-side. Note the `false` has `"tag": "Bool"`:
```
Serving JSON:
{
    ...,

    "_fooTestData": {
        "contents": {
            "contents": 9,
            "tag": "Int"
        },
        "tag": "Maybe"
    },
    "_fooTestSum": {
        "contents": false,
        "tag": "Bool"
    }
}
Received from client:
Foo message: Hola        Foo number: 124         Foo list length: 8
```

However, in the failing example above, the `false` inside the `Right` does not have `"tag": "Bool"` anywhere.

peterbecich commented:
Encoding of `Either` with `argonaut-aeson-generic` I think is broken.

peterbecich commented:
The issue is on the Haskell side. Aeson encoding of `Either` is incompatible with Purescript library `argonaut-aeson-generic`. Example:
```
    "_fooE": {
        "Left": "foooo"
    },
```
The Purescript library expects to find a `tag`.

peterbecich committed:
make Aeson produce `tag` for `Either`

peterbecich commented:
The `Either` has a `tag` now. Next issue is missing field `value`. The other datatypes decode fine with `contents`; why does `Either` need to have `value`?

```
Serving JSON:
{
    "_fooBaz": {
        "_bazMessage": "hello"
    },
    "_fooE": {
        "contents": 8,
        "tag": "Right"
    },
    "_fooList": [
        10,
        11,
        12,
        13
    ],
    "_fooMap": {
        "bar": 3,
        "baz": 3,
        "foo": 2
    },
    "_fooMessage": "Hello",
    "_fooNumber": 123,
    "_fooTestData": {
        "contents": "foooo",
        "tag": "TEither"
    },
    "_fooTestSum": {
        "contents": true,
        "tag": "Bool"
    }
}
```
```
Error decoding Foo: An error occurred while decoding a JSON value:
  Under 'When decoding a Foo':
  At object key '_fooE':
  Under 'Either':
  At object key 'value':
  No value was found.
```

peterbecich committed:
working example of encoding/decoding `Either`

peterbecich committed:
clean-up

peterbecich committed:
readme

peterbecich committed:
update purescript-bridge-json-helpers

https://github.com/input-output-hk/purescript-bridge-json-helpers/commits/main/

peterbecich committed:
update github action

peterbecich committed:
get tests passing

peterbecich committed:
improve readme

peterbecich committed:
improve readme, clean-up

peterbecich commented:
@eskimor, @kindaro, @bentongxyz , please review

I have updated the summary: https://github.com/eskimor/purescript-bridge/pull/89#issue-1910635911

IOHK's `RoundTrip` tests are the best check on this PR in my opinion. This are run in CI. `cabal test` will run them.

There are issues with Argonaut discussed here: https://github.com/purescript-contrib/purescript-argonaut-codecs/pull/116

There are likely to be simple issues in this PR like dead code, lint, etc. I think it is mostly solid.

flip111 commented:
What are your thoughts of using an approach like `codec-argonaut` instead of relying on type classes? As it was discussed here https://github.com/purescript-contrib/purescript-argonaut-codecs/issues/113#issuecomment-1656845390

One fork as you mentioned https://github.com/purescript-contrib/purescript-argonaut-codecs/issues/115#issuecomment-1874884848 can solve the issues of incompatibility with aeson. But that still leaves additional types on the table like `Ratio` from the first comment.

flip111 commented on stack.yaml.lock:
```diff
@@ -0,0 +1,12 @@
+- completed:
+    sha256: 1d663bec402f77fa09fa3e8cf288c3c7eb18a1f28a5e0c331ac47adeacd21346
+    size: 640034
+    url: https://raw.githubusercontent.com/commercialhaskell/stackage-snapshots/master/lts/21/16.yaml
```
Can we try LTS-22 yet?

peterbecich replied:
I tried to use GHC 9.6 on GitHub Actions and it failed, so I don't think LTS-22 will work

peterbecich commented:
@flip111 it sounds like a good idea to me. I don't think it will fundamentally change the PR. I think it will be an additional option to the three existing libraries.

kindaro commented:
As I see it, there are two big risks right now:

1. The biggest risk is that this patch does not get merged and all the work is lost to bit rot — to counteract this, it is better to merge sooner, even if something is not working.
2. The other, smaller risk is that merging future patches from the IOHK fork and other improvements will be hard — to counteract this risk, I think the patch should match some basic non-functional requirements like automatic formatting.

# review

## design, architecture, overall approach

This patch is of significant size and I cannot say anything meaningful as to its design and architecture. It would have been ideal if a series of smaller patches or atomic commits equipped with clear commit messages were offered, with each clearly marked as a feature or a breaking change. But I understand that making one big patch is faster. What matters now is that it gets merged.

## non-functional requirements

### good

* The code builds.
* There are no warnings even with `-Wall -Wextra`.
* I did not spot any dead code.

### bad

* There are 12 hints when running `hlint -XNoPatternSynonyms src`, 6 hints when running `hlint example` and 5 hints when running `hlint test`. In my ideal world, some of these hints are fixed and others are ignored by creating a file `hlint.yaml`.
* Formatting is not perfect. Running either `ormolu` or `stylish-haskell` does modify some files. In my ideal world, running `ormolu --mode inplace **/*.hs && stylish-haskell --inplace **/*.hs` has no effect because everything is already formatted this way. Also, it should be noted in documentation that `ormolu` should be used.
* I am unhappy about huge chunks of code commented out of `Printer.hs`. Do they serve any purpose? Can they simply be deleted?

## functional requirements

### good

* The working example shows that there is at least one way to use this library. So, we know this patch did not break the code altogether.

### bad

* There are some faults exposed by the new test suite. This is fine — so far as the foundation is strong, we can repair the local faults eventually.

# advice

My advice would be:

1. For @peterbecich to make `hlint` happy, remove code that is commented out _(if it needs to be put back eventually, make an issue about it)_ and automatically format the whole code base. This should take no more than a few hours.
2. For @flip111 to merge and consider adding some GitHub action that checks formatting and linting. In my ideal world, there is an interactive rebase step, but it would easily take a day of work so I understand if this does not happen.
3. In my ideal world, Peter and @flip111 manage to write a change log that explains what new features are added and what breaking changes are made in this patch.

# summary

Overall, this is great work. I shall be putting it into production as soon as my code base is upgraded to PureScript 0.15.

flip111 commented:
@kindaro thanks for the review. I care little for hlint and formatting, for me it's one of the lowest priorities. I understand the code review is difficult because of it's size, i thought the same. What would be good is to try a functional test. Take one or more of your own projects and make a new git branch if you will, then add peter's fork as a dependency. You can see [here](https://docs.haskellstack.org/en/stable/yaml_configuration/#extra-deps) how to do that with stack. Then observe the differences with the current library. Did you have to refactor your code to make it work again with the new version? Is it missing functionality? Are you hitting edge cases? Can you do a full round trip of serializing and deserializing. If you have `Arbitrary` together with `Aeson` on your data structures it would be useful to generate cases, serialize them, de-serialize them and then check them for equality with quickcheck or hedgehog.

kindaro commented:
> I care little for hlint and formatting, for me it's one of the lowest priorities.

This is of course up to you. I can only say that for me tidy code is easier to justify contributing to and depending on. What matters most right now is that the patch gets merged, and if linting and formatting are not blockers for you — the better!

> Take one or more of your own projects and make a new git branch if you will, then add peter's fork as a dependency.

This is exactly what I want to do, but unfortunately it is a bit complicated because the code I have right now does not work with the latest `purescript-bridge` and it will take significant work to get it to work. However, if something will not be working I am confident that I could fix it and send you a patch.

> If you have `Arbitrary` together with `Aeson` on your data structures it would be useful to generate cases, serialize them, de-serialize them and then check them for equality with quickcheck or hedgehog.

This is also something I should like to do, but at a later point in time. You should not wait for me on this.

peterbecich commented:
Thanks @kindaro , about formatting, `eskimor/purescript-bridge` is all Stylish Haskell; IOHK uses Ormolu. The repo originally had no consistent format chosen so I decided on Stylish Haskell and @eskimor agreed: https://github.com/eskimor/purescript-bridge/pull/76

There is a `.stylish-haskell` yaml. This should not produce any diffs:
```
find . -name '*.hs' | xargs stylish-haskell -i
```

I will respond to the other points later.

peterbecich committed:
clean-up

kindaro commented:
> `eskimor/purescript-bridge` is all Stylish Haskell; IOHK uses Ormolu.

Alright! I thought you eventually decided to format with `ormolu` to match the IOHK fork but I see now this is not so. By the way, formatting with `fourmolu` with the following configuration and then with `stylish-haskell` makes a diff of 787 lines, so we could make this standard if we wanted to have total formatting. Maybe it would help merge future changes from the IOHK fork?

```
indentation: 4
function-arrows: leading
comma-style: leading
import-export-style: leading
record-brace-space: true
haddock-style: multi-line
newlines-between-decls: 1
```

> This should not produce any diffs:
> 
> ```
> find . -name '*.hs' | xargs stylish-haskell -i
> ```

Alas:

```
% git status
On branch merge-iohk-3
Your branch is up to date with 'origin/merge-iohk-3'.

nothing to commit, working tree clean
% stylish-haskell --inplace **/*.hs
% git status
On branch merge-iohk-3
Your branch is up to date with 'origin/merge-iohk-3'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   src/Language/PureScript/Bridge/SumType.hs
	modified:   test/RoundTripArgonautAesonGeneric/Spec.hs
	modified:   test/RoundTripArgonautAesonGeneric/Types.hs
	modified:   test/RoundTripJsonHelpers/Spec.hs

no changes added to commit (use "git add" and/or "git commit -a")
```

peterbecich committed:
`nix flake update`

Replace NixPkgs `haskell-updates` branch with the default branch to simplify

peterbecich committed:
changelog

peterbecich committed:
stylish-haskell and lint

```
find . -name '*.hs' | xargs stylish-haskell -i
```

peterbecich commented:
>  By the way, formatting with fourmolu with the following settings and then with stylish-haskell makes a diff of 814 lines, compared with 4186 lines if ormolu is used, so we could make these settings standard if we wanted to have total formatting. Maybe it would help merge future changes from the IOHK fork?

I did something like this here: https://github.com/eskimor/purescript-bridge/pull/89/commits/b5fbac0dbdfac82e230d46082a895c49bbb9aed7
The best results at reversing IOHK's Ormolu formatting were to re-format first with Fourmolu, then stylish-haskell. Directly formatting IOHK with stylish-haskell did not work as well.

I agree there could have been something missed in the IOHK fork which could be merged upstream later. The easiest way to proceed there i.m.o. is to format IOHK with Fourmolu, then stylish-haskell.

kindaro commented:
Yes, thank you for putting my mind to the right spot. I was remembering that you formatted everything some time ago, but I was confused between `ormolu` and `fourmolu`. So, the code base has perfect formatting at commit b5fbac0dbdfac82e230d46082a895c49bbb9aed7. What if we were to rebase and format every subsequent commit in the same way? If this sounds like a lot of work, I have some scripts that can do it in 1 minute.

I come with better news though. I came around to try this patch on an industrial grade code base, and at once I found a fault.

# review

## functional requirements

### bad

* This patch breaks the feature I added in #85. I see you dropped it in favour of code from the IOHK fork, but their code generates broken instances for types with phantom parameters. Here is an example:

  ```haskell
  newtype ID a = ID {getID :: Int}
  ````

  The code form #85 generates this:

  ```PureScript
  newtype ID a =
      ID {
        getID :: Int
      }

  instance showID ∷ Show (ID a) where
  show value = genericShow value
  ```

  However, the code from this patch — presumably coming from IOHK — generates this:

  ```PureScript
  newtype ID a = ID { getID :: Int }

  instance (Show a) => Show (ID a) where
    show a = genericShow a
  ````

  Notice the spurious constraint `Show a`. This will break compilation of other types that depend on `ID` and want to have a `Show` of their own — they should not need an instance of `Show` for whichever phantom variable they instantiate `ID` with.

# summary

Generating healthy instances when phantom types are at play is important for my use cases. This is certainly not something I could not fix, but maybe we can try not breaking it in the first place?

kindaro commented:
By the way, the current `master` of `purescript-bridge` can handle `data` declarations with phantom type parameters, but not with non-phantom type parameters. The issue seems to be that the fake types like `A` that `purescript-bridge` instantiates type variables with do not have a `Show` instance of their own. I do not know exactly how this is supposed to work. It seems, however, that `Eq` and `Ord` for `data` declarations with non-phantom type parameters also do not work. Is this something this patch improves upon?

peterbecich commented:
Excellent points, thanks @kindaro , I had not noticed these issues. Thinking about how to fix.

flip111 commented:
> Should I squash my commits?

Let them be as tribute to your work :wink:

kindaro commented:
@peterbecich   I pray for your good health.

peterbecich committed:
demonstrate `json-helpers` in the `example`

peterbecich committed:
remove `Show` constraint; there are issues to be fixed

try example

https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1890994859

peterbecich committed:
example, generate PureScript before server starts

peterbecich commented:
I agree this should be fixed: https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1890994859
Here is the easy part of the change: https://github.com/peterbecich/purescript-bridge/commit/e663a3d31c18e9474e0da1818234e01894132dad
However, this breaks other things which I haven't had time to look into.

The `example` now covers `json-helpers`; if you run `example` on `merge-iohk-3`, it works. If you run `example` on branch `fix-show-on-iohk-fork`, it will show the issues with this fix remaining to be solved.

flip111 commented:
# Questions
* Why were switches removed?

# Stuff that i noticed that i liked
* Added word8Bridge and word16Bridge
* Added `Eq A`

# Problems

## Should not derive `Enum` and `Bounded` when `Ord` is not specified and missing `Show` instance

haskell in
```haskell
data WeekInMonth = WeekFirst | WeekSecond | WeekThird | WeekFourth | WeekLast
  deriving (Generic, Eq, Show)
  deriving TextShow via FromGeneric WeekInMonth
instance Aeson.ToJSON WeekInMonth where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions
instance Aeson.FromJSON WeekInMonth
```

purescript out
```purescript
data WeekInMonth
  = WeekFirst
  | WeekSecond
  | WeekThird
  | WeekFourth
  | WeekLast

derive instance Eq WeekInMonth

derive instance Generic WeekInMonth _

instance Enum WeekInMonth where
  succ = genericSucc
  pred = genericPred

instance Bounded WeekInMonth where
  bottom = genericBottom
  top = genericTop
```

Problem description
* `Enum` requires `Ord` but it is not available. Solution: should not generate `Enum` when `Ord` is not there. Possibly could give an info message that `Enum` was skipped
* Save as above but then for `Bounded`
* specified `Show` instance is missing
* specified `Aeson` instances missing

Previously code looked like this
```haskell
data WeekInMonth =
    WeekFirst
  | WeekSecond
  | WeekThird
  | WeekFourth
  | WeekLast

derive instance eqWeekInMonth :: Eq WeekInMonth
instance showWeekInMonth :: Show WeekInMonth where
  show = genericShow
instance encodeJsonWeekInMonth :: EncodeJson WeekInMonth where
  encodeJson = genericEncodeAeson Argonaut.defaultOptions
instance decodeJsonWeekInMonth :: DecodeJson WeekInMonth where
  decodeJson = genericDecodeAeson Argonaut.defaultOptions
derive instance genericWeekInMonth :: Generic WeekInMonth _
```

flip111 commented:
@peterbecich do you have a list of open points that still need work?

peterbecich commented:
@flip111 thank you for the detailed comment, I will read carefully later

One thing I need help with is fixing the `generic Show` instance. It was correctly implemented by @kindaro, then it was broken in my PR: https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1890994859
Here is how far I got with that: https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1913755047
You can use the `example` to check it.

flip111 committed:
Fix show instances for unused type variables

peterbecich commented:
Thank you @flip111 : https://github.com/peterbecich/purescript-bridge/pull/1

peterbecich committed:
complete upgrade to new Spago@next

flip111 committed:
move example to new Spago

`spago migrate`

peterbecich committed:
Merge pull request #2 from peterbecich/spago-upgrade

Spago upgrade

peterbecich committed:
update example package set

flip111 committed:
Revert eta reduction of show = genericShow

peterbecich committed:
Merge pull request #1 from flip111/fix-show2

Fix show instances for unused type variables

peterbecich committed:
extra-source-files

peterbecich commented:
TODO `Enum` and `Ord` https://github.com/peterbecich/purescript-bridge/pull/1#issue-2137164432

peterbecich committed:
github action tweak

flip111 commented:
@peterbecich did you look at the problem of `Enum` and `Ord`? Can you confirm or otherwise comment whether the findings here are correct https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1934932243 ?

peterbecich committed:
`nix flake update`

peterbecich committed:
calendar datatype test

https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1934932243

peterbecich committed:
use quasiquoter to clean up tests

https://hackage.haskell.org/package/neat-interpolation-0.5.1.4/docs/NeatInterpolation.html

https://github.com/ghc-proposals/ghc-proposals/issues/260

peterbecich committed:
`Enum` and `Bounded` require `Eq` and `Ord`

https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1934932243

peterbecich commented:
@flip111 , in progress, thanks

flip111 commented:
@peterbecich thanks. Need any more help?

lucasdicioccio committed:
Support GHC.Tuple.Prim.Tuple2

peterbecich committed:
`nix flake update`, update Spago

peterbecich committed:
use Spago Workspaces

https://github.com/purescript/spago?tab=readme-ov-file#monorepo-support
https://github.com/purescript/spago?tab=readme-ov-file#the-workspace

peterbecich committed:
provide esbuild

peterbecich committed:
`nix flake update`

peterbecich committed:
example List type

peterbecich committed:
fix bridge from Haskell List to PureScript Array for GHC 9.6

peterbecich committed:
improve GitHub Action

peterbecich committed:
`nix flake update`

peterbecich commented:
Once https://github.com/purescript/spago/pull/1310 is merged and Spago is updated via `nix flake update`, the test suite should be closer to working.  Right now, `cabal test` fails because Spago cannot find the `workspace` at the root of the project.

peterbecich committed:
Merge remote-tracking branch 'parent/master' into merge-iohk-3

flip111 commented:
@kindaro @peterbecich do you also find that for newer GHC releases `import GHC.Tuple.Prim (Tuple2)` is being used instead of `import Data.Tuple (Tuple)`? I'm using [LTS 23.9 for ghc-9.8.4](https://www.stackage.org/lts-23.9)

peterbecich committed:
update Nix Flake

get Spago 0.93.44

use ghc 9.8

update spago.lock

peterbecich committed:
provide Eq and Ord in example

peterbecich committed:
fix one test

peterbecich committed:
add tuple example

peterbecich committed:
fix duplicate instance printing; fix Unit printing

peterbecich committed:
fix Round Trip integration tests

peterbecich committed:
fix GitHub Actions

replace Magic Nix Cache with https://github.com/nix-community/cache-nix-action

peterbecich committed:
fix compilation of example

peterbecich committed:
update Stack build, provide GitHub Action coverage for Stack

peterbecich committed:
add tests to Nix Flake GitHub Action

peterbecich commented:
* `Either` works great https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1769345856 with the `json-helpers` library https://github.com/input-output-hk/purescript-bridge-json-helpers
* All of the example types work correctly with `json-helpers`
* Issues like this https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-1934932243 are unsolved
  * This should be improved in another PR
  * I don't think it is essential to improve now because the library fundamentally works
  * Tests and example provide excellent coverage for this PR 
* https://github.com/eskimor/purescript-bridge/pull/89#issuecomment-2663330072 all issues with `Tuple2` have been solved
  * it works in all builds (Cabal & Stack); there is GitHub Actions coverage
* there are problems with `Tuple3+` which should be solved in follow-up PR

peterbecich commented:
@eskimor , @flip111 , @kindaro , please review

This branch is in a working state. Test and example coverage is excellent. This PR should not be a regression.

The library is easier to iterate and test now. For any issues that still exist, it should be easier for myself and others to follow up on them.

peterbecich committed:
update development.md
